src/main/java/com/google/cloud/spanner/r2dbc/SpannerStatement.java

```diff
@@ -55,45 +74,84 @@
   public SpannerStatement(
       Client client,
       Session session,
-      @Nullable Transaction transaction,
+      @Nullable SpannerTransactionContext transaction,
       String sql) {
 
     this.client = client;
     this.session = session;
     this.transaction = transaction;
-    this.sql = sql;
+    this.sql = requireNonNull(sql, "SQL string can not be null");
   }
 
   @Override
   public Statement add() {
-    return null;
+    if (this.currentBindingsBuilder != null) {
+      this.bindingsStucts.add(this.currentBindingsBuilder.build());
+    }
+    this.currentBindingsBuilder = null;
+    return this;
   }
 
   @Override
-  public Statement bind(Object o, Object o1) {
-    return null;
+  public Statement bind(Object identifier, Object value) {
+    requireNonNull(identifier);
+    if (identifier instanceof String) {
+      String paramName = (String) identifier;
+      //we assume all parameters with the same name have the same type
```

Comment:
style nit: space before start of comment text
```suggestion
      // we assume all parameters with the same name have the same type
```